File: backend/app/models/conversation.py
"""
Conversation model representing chat messages between customers and AI.
"""

from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field


class Conversation(BaseModel):
    """
    Conversation model representing a chat message exchange.

    Attributes:
        id: Unique conversation identifier (UUID)
        user_id: Store owner UUID
        customer_identifier: Anonymous customer ID (email, session, etc.)
        customer_message: Customer's question or message
        ai_response: AI-generated response
        message_count: Sequential message number in conversation
        products_referenced: List of product IDs mentioned in response
        intent_detected: Detected customer intent
        sentiment_score: Customer sentiment (-1.0 to 1.0)
        response_time_ms: AI response time in milliseconds
        created_at: Message timestamp
    """

    id: Optional[str] = Field(None, description="Conversation UUID")
    user_id: str = Field(..., description="Store owner UUID")
    customer_identifier: Optional[str] = Field(
        None, max_length=255, description="Customer identifier"
    )
    customer_message: str = Field(..., description="Customer message")
    ai_response: str = Field(..., description="AI response")
    message_count: int = Field(default=1, ge=1, description="Message sequence number")
    products_referenced: List[str] = Field(
        default_factory=list, description="Product IDs mentioned"
    )
    intent_detected: Optional[str] = Field(
        None, max_length=100, description="Detected intent"
    )
    sentiment_score: Optional[float] = Field(
        None, ge=-1.0, le=1.0, description="Sentiment score"
    )
    response_time_ms: Optional[int] = Field(
        None, ge=0, description="Response time in ms"
    )
    created_at: Optional[datetime] = Field(None, description="Message timestamp")

    class Config:
        """Pydantic configuration"""

        json_schema_extra = {
            "example": {
                "id": "123e4567-e89b-12d3-a456-426614174000",
                "user_id": "123e4567-e89b-12d3-a456-426614174001",
                "customer_identifier": "customer@example.com",
                "customer_message": "Do you have wireless headphones?",
                "ai_response": "Yes! We have several wireless headphones available...",
                "message_count": 1,
                "products_referenced": ["123e4567-e89b-12d3-a456-426614174002"],
                "intent_detected": "product_inquiry",
                "sentiment_score": 0.8,
                "response_time_ms": 1523,
            }
        }

    def to_dict(self) -> Dict[str, Any]:
        """
        Convert model to dictionary.

        Returns:
            Dictionary representation
        """
        return self.model_dump()

    def to_chat_dict(self) -> Dict[str, Any]:
        """
        Convert to chat display format.

        Returns:
            Dictionary formatted for chat UI
        """
        return {
            "id": self.id,
            "message": self.customer_message,
            "response": self.ai_response,
            "timestamp": self.created_at.isoformat() if self.created_at else None,
            "products": self.products_referenced,
        }


class ConversationCreate(BaseModel):
    """Schema for creating a new conversation message"""

    customer_identifier: Optional[str] = Field(None, max_length=255)
    customer_message: str = Field(...)
    ai_response: str = Field(...)
    message_count: int = Field(default=1, ge=1)
    products_referenced: List[str] = Field(default_factory=list)
    intent_detected: Optional[str] = Field(None, max_length=100)
    sentiment_score: Optional[float] = Field(None, ge=-1.0, le=1.0)
    response_time_ms: Optional[int] = Field(None, ge=0)


class ConversationHistory(BaseModel):
    """Schema for conversation history response"""

    conversations: List[Conversation] = Field(default_factory=list)
    total_count: int = Field(default=0)
    page: int = Field(default=1)
    page_size: int = Field(default=20)
    has_more: bool = Field(default=False)

    class Config:
        """Pydantic configuration"""

        json_schema_extra = {
            "example": {
                "conversations": [],
                "total_count": 47,
                "page": 1,
                "page_size": 20,
                "has_more": True,
            }
        }


class MessageIntent:
    """Common message intent types"""

    PRODUCT_INQUIRY = "product_inquiry"
    ORDER_LOOKUP = "order_lookup"
    GENERAL_QUESTION = "general_question"
    COMPLAINT = "complaint"
    PRAISE = "praise"
    SHIPPING_QUESTION = "shipping_question"
    RETURN_REQUEST = "return_request"
    UNKNOWN = "unknown"


__all__ = [
    "Conversation",
    "ConversationCreate",
    "ConversationHistory",
    "MessageIntent",
]

File: flutter_app/lib/models/conversation.dart
import 'package:hive/hive.dart';

part 'conversation.g.dart';

/// Conversation model representing chat messages
@HiveType(typeId: 0)
class Conversation {
  @HiveField(0)
  final String id;

  @HiveField(1)
  final String userId;

  @HiveField(2)
  final String? customerIdentifier;

  @HiveField(3)
  final String customerMessage;

  @HiveField(4)
  final String aiResponse;

  @HiveField(5)
  final int messageCount;

  @HiveField(6)
  final List<String> productsReferenced;

  @HiveField(7)
  final String? intentDetected;

  @HiveField(8)
  final double? sentimentScore;

  @HiveField(9)
  final int? responseTimeMs;

  @HiveField(10)
  final DateTime createdAt;

  Conversation({
    required this.id,
    required this.userId,
    this.customerIdentifier,
    required this.customerMessage,
    required this.aiResponse,
    this.messageCount = 1,
    this.productsReferenced = const [],
    this.intentDetected,
    this.sentimentScore,
    this.responseTimeMs,
    required this.createdAt,
  });

  factory Conversation.fromJson(Map<String, dynamic> json) {
    return Conversation(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      customerIdentifier: json['customer_identifier'] as String?,
      customerMessage: json['customer_message'] as String,
      aiResponse: json['ai_response'] as String,
      messageCount: json['message_count'] as int? ?? 1,
      productsReferenced: (json['products_referenced'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      intentDetected: json['intent_detected'] as String?,
      sentimentScore: json['sentiment_score'] as double?,
      responseTimeMs: json['response_time_ms'] as int?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'customer_identifier': customerIdentifier,
      'customer_message': customerMessage,
      'ai_response': aiResponse,
      'message_count': messageCount,
      'products_referenced': productsReferenced,
      'intent_detected': intentDetected,
      'sentiment_score': sentimentScore,
      'response_time_ms': responseTimeMs,
      'created_at': createdAt.toIso8601String(),
    };
  }

  /// Convert to chat display format
  Map<String, dynamic> toChatJson() {
    return {
      'id': id,
      'message': customerMessage,
      'response': aiResponse,
      'timestamp': createdAt.toIso8601String(),
      'products': productsReferenced,
    };
  }

  Conversation copyWith({
    String? id,
    String? userId,
    String? customerIdentifier,
    String? customerMessage,
    String? aiResponse,
    int? messageCount,
    List<String>? productsReferenced,
    String? intentDetected,
    double? sentimentScore,
    int? responseTimeMs,
    DateTime? createdAt,
  }) {
    return Conversation(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      customerIdentifier: customerIdentifier ?? this.customerIdentifier,
      customerMessage: customerMessage ?? this.customerMessage,
      aiResponse: aiResponse ?? this.aiResponse,
      messageCount: messageCount ?? this.messageCount,
      productsReferenced: productsReferenced ?? this.productsReferenced,
      intentDetected: intentDetected ?? this.intentDetected,
      sentimentScore: sentimentScore ?? this.sentimentScore,
      responseTimeMs: responseTimeMs ?? this.responseTimeMs,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  @override
  String toString() {
    return 'Conversation(id: $id, message: ${customerMessage.substring(0, 20)}...)';
  }
}

/// Message intent types
class MessageIntent {
  static const String productInquiry = 'product_inquiry';
  static const String orderLookup = 'order_lookup';
  static const String generalQuestion = 'general_question';
  static const String complaint = 'complaint';
  static const String praise = 'praise';
  static const String shippingQuestion = 'shipping_question';
  static const String returnRequest = 'return_request';
  static const String unknown = 'unknown';

  static String getDisplayName(String intent) {
    switch (intent) {
      case productInquiry:
        return 'Product Question';
      case orderLookup:
        return 'Order Lookup';
      case generalQuestion:
        return 'General Question';
      case complaint:
        return 'Complaint';
      case praise:
        return 'Praise';
      case shippingQuestion:
        return 'Shipping Question';
      case returnRequest:
        return 'Return Request';
      default:
        return 'Unknown';
    }
  }
}

/// Chat message request
class ChatMessageRequest {
  final String message;
  final String? customerIdentifier;
  final String? sessionId;

  ChatMessageRequest({
    required this.message,
    this.customerIdentifier,
    this.sessionId,
  });

  Map<String, dynamic> toJson() {
    return {
      'message': message,
      'customer_identifier': customerIdentifier,
      'session_id': sessionId,
    };
  }
}

/// Chat message response
class ChatMessageResponse {
  final String id;
  final String customerMessage;
  final String aiResponse;
  final List<Map<String, dynamic>> products;
  final String? intentDetected;
  final int responseTimeMs;

  ChatMessageResponse({
    required this.id,
    required this.customerMessage,
    required this.aiResponse,
    this.products = const [],
    this.intentDetected,
    required this.responseTimeMs,
  });

  factory ChatMessageResponse.fromJson(Map<String, dynamic> json) {
    return ChatMessageResponse(
      id: json['id'] as String,
      customerMessage: json['customer_message'] as String,
      aiResponse: json['ai_response'] as String,
      products: (json['products'] as List<dynamic>?)
              ?.map((e) => e as Map<String, dynamic>)
              .toList() ??
          [],
      intentDetected: json['intent_detected'] as String?,
      responseTimeMs: json['response_time_ms'] as int,
    );
  }
}
